pkg/synchronizer/opstate.go

```diff
@@ -0,0 +1,198 @@
+	sim := s.getSimCardByImsi(site, imsi)
+	if sim == nil {
+		return nil
+	}
+	for _, dev := range site.Device {
+		if *dev.SimCard == *sim.SimId {
+			return dev
+		}
+	}
+	return nil
+}
+
+// get a prometheus metrics fetcher, caching previously-created fetchers
+func (s *Synchronizer) getFetcher(endpoint string) (*metrics.Fetcher, error) {
+	mf, okay := s.prometheus[endpoint]
+	if okay {
+		return mf, nil
+	}
+
+	mf, err := metrics.NewFetcher(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	s.prometheus[endpoint] = mf
+	return mf, nil
+}
+
+// update Device.IPAddress opstate
+func (s *Synchronizer) updateDeviceIPAddress(config *gnmi.ConfigForest, entID string, siteID string, imsi uint64, mobileIP string, connected float64) {
+	// Hold the lock while updating the config tree
+	config.Mu.Lock()
+	defer config.Mu.Unlock()
+
+	entRoot, okay := config.Configs[entID]
+	if !okay {
+		// Rare corner case -- the enterprise has disappeared. Maybe some other thread deleted it
+		log.Warnf("Failed to get enterprise %s", entID)
+		return
+	}
+	enterprise := entRoot.(*RootDevice)
+
+	site, okay := enterprise.Site[siteID]
+	if !okay {
+		// Rare corner case -- the site has disappeared. Maybe some other thread deleted it
+		log.Warnf("Failed to get site %s", siteID)
+		return
+	}
+
+	device := s.getDeviceByImsi(site, imsi)
+	if device == nil {
+		// Can happen if the device is reported in prometheus, but does not exist in
+		// ROC. Rare in real life, but plausible in test infrastructure.
+		// Message is currently disabled, as it is fairly noisy.
+		//log.Warnf("Failed to find device %v in site %s of enterprise %s", imsi, siteID, entID)
+		return
+	}
+
+	if device.State == nil {
+		device.State = &DeviceState{}
+	}
+	device.State.IpAddress = aStr(mobileIP)
+
+	if connected != 0 {
+		device.State.Connected = aStr("Yes")
+	} else {
+		device.State.Connected = aStr("No")
+	}
+}
+
+// poll promtheus for metrics, and add them to the config tree
+func (s *Synchronizer) pollPrometheus(config *gnmi.ConfigForest, prom *prometheusInfo) error {
+	log.Debugf("Polling prometheus site %s enterprise %s endpoint %s", prom.siteID, prom.entID, prom.endpoint)
+
+	fetcher, err := s.getFetcher(prom.endpoint)
+	if err != nil {
+		return err
+	}
+	v, err := fetcher.GetVector("subscribers_info")
+	if err != nil {
+		return err
+	}
+
+	// vector is []*Sample
+	// sample is {Metric Value:SampleValue Time}
+	// metric is a labelset
+	// labelsite is map[labelname]labelvalue
+	// samplevalue is float64
+
+	for _, sample := range v {
+		// labels will be imsi and mobile_ip
+		imsiString, okay := sample.Metric["imsi"]
+		if !okay {
+			log.Warn("Failed to get imsi from sample")
+			continue
+		}
+		mobileIP, okay := sample.Metric["mobile_ip"]
+		if !okay {
+			log.Warn("Failed to get mobile_ip from sample")
+			continue
+		}
+
+		imsi, err := strconv.ParseUint(string(imsiString), 10, 64)
+		if err != nil {
+			log.Warn("Failed to convert imsi from sample")
```

Comment:
it would be good to log the errant value

Reply:
Will do.